From 4b9aa65c2aefb3cf935c9be2628cd2fd10789364 Mon Sep 17 00:00:00 2001
From: "Petar G. Georgiev" <petarg@codeaurora.org>
Date: Tue, 27 Aug 2019 14:11:17 +0300
Subject: [CHANGE 4/4] waylandsink: support gbm buffer backend protocol

Support the wayland gbm_buffer_backend protocol.
SHM formats, DMABuf formats and GBM formats are exposed differently
in caps: the GBM formats are flagged with GST_CAPS_FEATURE_MEMORY_GBM.
No buffer pool is proposed for GBM buffers, it is the upstream
element responsibility to provide with such buffers
---
 ext/wayland/Makefile.am      |   8 ++-
 ext/wayland/gstwaylandsink.c |  38 +++++++++--
 ext/wayland/gstwaylandsink.h |   1 +
 ext/wayland/meson.build      |   2 +
 ext/wayland/wldisplay.c      |  63 ++++++++++++++++-
 ext/wayland/wldisplay.h      |  10 +++
 ext/wayland/wlgbmallocator.c | 159 +++++++++++++++++++++++++++++++++++++++++++
 ext/wayland/wlgbmallocator.h |  38 +++++++++++
 ext/wayland/wlvideoformat.c  | 110 +++++++++++++++++++++++-------
 ext/wayland/wlvideoformat.h  |   5 +-
 10 files changed, 401 insertions(+), 33 deletions(-)
 create mode 100644 ext/wayland/wlgbmallocator.c
 create mode 100644 ext/wayland/wlgbmallocator.h

diff --git a/ext/wayland/Makefile.am b/ext/wayland/Makefile.am
index 854ad1f..715c401 100644
--- a/ext/wayland/Makefile.am
+++ b/ext/wayland/Makefile.am
@@ -3,6 +3,8 @@ plugin_LTLIBRARIES = libgstwaylandsink.la
 BUILT_SOURCES = \
 	viewporter-protocol.c \
 	viewporter-client-protocol.h \
+	gbm-buffer-backend-protocol.c \
+	gbm-buffer-backend-client-protocol.h \
 	linux-dmabuf-unstable-v1-protocol.c \
 	linux-dmabuf-unstable-v1-client-protocol.h \
 	xdg-shell-protocol.c \
@@ -13,6 +15,7 @@ BUILT_SOURCES = \
 libgstwaylandsink_la_SOURCES =  \
 	gstwaylandsink.c \
 	wlshmallocator.c \
+	wlgbmallocator.c \
 	wlbuffer.c \
 	wldisplay.c \
 	wlwindow.c \
@@ -21,6 +24,7 @@ libgstwaylandsink_la_SOURCES =  \
 
 nodist_libgstwaylandsink_la_SOURCES = \
 	viewporter-protocol.c \
+	gbm-buffer-backend-protocol.c \
 	linux-dmabuf-unstable-v1-protocol.c \
 	xdg-shell-protocol.c \
 	scaler-protocol.c
@@ -35,12 +39,14 @@ libgstwaylandsink_la_LIBADD = \
 	$(GST_PLUGINS_BASE_LIBS) \
 	-lgstvideo-$(GST_API_VERSION) \
 	-lgstallocators-$(GST_API_VERSION) \
-	$(WAYLAND_LIBS)
+	$(WAYLAND_LIBS) \
+	-lgbm
 libgstwaylandsink_la_LDFLAGS = $(GST_PLUGIN_LDFLAGS)
 
 noinst_HEADERS = \
 	gstwaylandsink.h \
 	wlshmallocator.h \
+	wlgbmallocator.h \
 	wlbuffer.h \
 	wldisplay.h \
 	wlwindow.h \
diff --git a/ext/wayland/gstwaylandsink.c b/ext/wayland/gstwaylandsink.c
index 553a3b4..9bf51b3 100644
--- a/ext/wayland/gstwaylandsink.c
+++ b/ext/wayland/gstwaylandsink.c
@@ -46,6 +46,7 @@
 #include "wlbuffer.h"
 #include "wlshmallocator.h"
 #include "wllinuxdmabuf.h"
+#include "wlgbmallocator.h"
 
 #include <gst/wayland/wayland.h>
 #include <gst/video/videooverlay.h>
@@ -83,6 +84,8 @@ static GstStaticPadTemplate sink_template = GST_STATIC_PAD_TEMPLATE ("sink",
     GST_PAD_ALWAYS,
     GST_STATIC_CAPS (GST_VIDEO_CAPS_MAKE (WL_VIDEO_FORMATS) ";"
         GST_VIDEO_CAPS_MAKE_WITH_FEATURES (GST_CAPS_FEATURE_MEMORY_DMABUF,
+            WL_VIDEO_FORMATS) ";"
+        GST_VIDEO_CAPS_MAKE_WITH_FEATURES (GST_CAPS_FEATURE_MEMORY_GBM,
             WL_VIDEO_FORMATS))
     );
 
@@ -533,19 +536,20 @@ gst_wayland_sink_get_caps (GstBaseSink * bsink, GstCaps * filter)
 
   if (sink->display) {
     GValue shm_list = G_VALUE_INIT, dmabuf_list = G_VALUE_INIT;
-    GValue value = G_VALUE_INIT;
+    GValue gbm_list = G_VALUE_INIT, value = G_VALUE_INIT;
     GArray *formats;
     gint i;
     guint fmt;
 
     g_value_init (&shm_list, GST_TYPE_LIST);
     g_value_init (&dmabuf_list, GST_TYPE_LIST);
+    g_value_init (&gbm_list, GST_TYPE_LIST);
 
     /* Add corresponding shm formats */
     formats = sink->display->shm_formats;
     for (i = 0; i < formats->len; i++) {
       g_value_init (&value, G_TYPE_STRING);
-      fmt = g_array_index (formats, uint32_t, i);
+      fmt = g_array_index (formats, guint, i);
       g_value_set_static_string (&value, gst_wl_shm_format_to_string (fmt));
       gst_value_list_append_and_take_value (&shm_list, &value);
     }
@@ -557,7 +561,7 @@ gst_wayland_sink_get_caps (GstBaseSink * bsink, GstCaps * filter)
     formats = sink->display->dmabuf_formats;
     for (i = 0; i < formats->len; i++) {
       g_value_init (&value, G_TYPE_STRING);
-      fmt = g_array_index (formats, uint32_t, i);
+      fmt = g_array_index (formats, guint, i);
       g_value_set_static_string (&value, gst_wl_dmabuf_format_to_string (fmt));
       gst_value_list_append_and_take_value (&dmabuf_list, &value);
     }
@@ -565,6 +569,17 @@ gst_wayland_sink_get_caps (GstBaseSink * bsink, GstCaps * filter)
     gst_structure_take_value (gst_caps_get_structure (caps, 1), "format",
         &dmabuf_list);
 
+    /* Add corresponding gbm formats */
+    formats = sink->display->gbm_formats;
+    for (i = 0; i < formats->len; i++) {
+      g_value_init (&value, G_TYPE_STRING);
+      fmt = g_array_index (formats, guint, i);
+      g_value_set_static_string (&value, gst_wl_gbm_format_to_string (fmt));
+      gst_value_list_append_and_take_value (&gbm_list, &value);
+    }
+
+    gst_structure_take_value (gst_caps_get_structure (caps, 2), "format",
+        &gbm_list);
     GST_DEBUG_OBJECT (sink, "display caps: %" GST_PTR_FORMAT, caps);
   }
 
@@ -610,7 +625,7 @@ static gboolean
 gst_wayland_sink_set_caps (GstBaseSink * bsink, GstCaps * caps)
 {
   GstWaylandSink *sink;
-  gboolean use_dmabuf;
+  gboolean use_dmabuf, use_gbm;
   GstVideoFormat format;
 
   sink = GST_WAYLAND_SINK (bsink);
@@ -632,15 +647,22 @@ gst_wayland_sink_set_caps (GstBaseSink * bsink, GstCaps * caps)
   use_dmabuf = gst_caps_features_contains (gst_caps_get_features (caps, 0),
       GST_CAPS_FEATURE_MEMORY_DMABUF);
 
+  use_gbm = gst_caps_features_contains (gst_caps_get_features (caps, 0),
+      GST_CAPS_FEATURE_MEMORY_GBM);
+
   /* validate the format base on the memory type. */
   if (use_dmabuf) {
     if (!gst_wl_display_check_format_for_dmabuf (sink->display, format))
       goto unsupported_format;
+  } else if (use_gbm) {
+    if (!gst_wl_display_check_format_for_gbm (sink->display, format))
+      goto unsupported_format;
   } else if (!gst_wl_display_check_format_for_shm (sink->display, format)) {
     goto unsupported_format;
   }
 
   sink->use_dmabuf = use_dmabuf;
+  sink->use_gbm = use_gbm;
 
   return TRUE;
 
@@ -685,7 +707,7 @@ gst_wayland_sink_propose_allocation (GstBaseSink * bsink, GstQuery * query)
 }
 
 static void
-frame_redraw_callback (void *data, struct wl_callback *callback, uint32_t time)
+frame_redraw_callback (void *data, struct wl_callback *callback, guint time)
 {
   GstWaylandSink *sink = data;
 
@@ -722,6 +744,7 @@ render_last_buffer (GstWaylandSink * sink, gboolean redraw)
     info = &sink->video_info;
     sink->video_info_changed = FALSE;
   }
+
   gst_wl_window_render (sink->window, wlbuffer, info);
 }
 
@@ -806,6 +829,10 @@ gst_wayland_sink_show_frame (GstVideoSink * vsink, GstBuffer * buffer)
     if (nb_dmabuf && (nb_dmabuf == gst_buffer_n_memory (buffer)))
       wbuf = gst_wl_linux_dmabuf_construct_wl_buffer (buffer, sink->display,
           &sink->video_info);
+  } else if (gst_wl_display_check_format_for_gbm (sink->display, format)) {
+    if (gst_buffer_n_memory (buffer) == 1 && gst_is_fd_memory (mem))
+      wbuf = gst_wl_gbm_memory_construct_wl_buffer (buffer, sink->display,
+          &sink->video_info);
   }
 
   if (!wbuf && gst_wl_display_check_format_for_shm (sink->display, format)) {
@@ -902,6 +929,7 @@ render:
 
   if (buffer != to_render)
     gst_buffer_unref (to_render);
+
   goto done;
 
 no_window_size:
diff --git a/ext/wayland/gstwaylandsink.h b/ext/wayland/gstwaylandsink.h
index 3083967..b2e1618 100644
--- a/ext/wayland/gstwaylandsink.h
+++ b/ext/wayland/gstwaylandsink.h
@@ -57,6 +57,7 @@ struct _GstWaylandSink
   GstWlWindow *window;
   GstBufferPool *pool;
   gboolean use_dmabuf;
+  gboolean use_gbm;
 
   gboolean video_info_changed;
   GstVideoInfo video_info;
diff --git a/ext/wayland/meson.build b/ext/wayland/meson.build
index 654a644..49cd2c7 100644
--- a/ext/wayland/meson.build
+++ b/ext/wayland/meson.build
@@ -15,6 +15,8 @@ if use_wayland
         ['/stable/viewporter/viewporter.xml', 'viewporter-protocol.c', 'viewporter-client-protocol.h'],
         ['/unstable/linux-dmabuf/linux-dmabuf-unstable-v1.xml',
          'linux-dmabuf-unstable-v1-protocol.c', 'linux-dmabuf-unstable-v1-client-protocol.h']
+        ['/stable/gbm-buffer-backend/gbm-buffer-backend.xml' 'gbm-buffer-backend-protocol.c'
+         'gbm-buffer-backend-client-protocol.h']
     ]
     protocols_files = []
 
diff --git a/ext/wayland/wldisplay.c b/ext/wayland/wldisplay.c
index 4aa0b3e..70d647b 100644
--- a/ext/wayland/wldisplay.c
+++ b/ext/wayland/wldisplay.c
@@ -27,6 +27,8 @@
 #include "wlvideoformat.h"
 
 #include <errno.h>
+#include <fcntl.h>
+#include <unistd.h>
 
 GST_DEBUG_CATEGORY_EXTERN (gstwayland_debug);
 #define GST_CAT_DEFAULT gstwayland_debug
@@ -47,6 +49,7 @@ gst_wl_display_init (GstWlDisplay * self)
 {
   self->shm_formats = g_array_new (FALSE, FALSE, sizeof (uint32_t));
   self->dmabuf_formats = g_array_new (FALSE, FALSE, sizeof (uint32_t));
+  self->gbm_formats = g_array_new (FALSE, FALSE, sizeof (uint32_t));
   self->wl_fd_poll = gst_poll_new (TRUE);
   self->buffers = g_hash_table_new (g_direct_hash, g_direct_equal);
   g_mutex_init (&self->buffers_mutex);
@@ -74,6 +77,7 @@ gst_wl_display_finalize (GObject * gobject)
 
   g_array_unref (self->shm_formats);
   g_array_unref (self->dmabuf_formats);
+  g_array_unref (self->gbm_formats);
   gst_poll_free (self->wl_fd_poll);
   g_hash_table_unref (self->buffers);
   g_mutex_clear (&self->buffers_mutex);
@@ -90,6 +94,9 @@ gst_wl_display_finalize (GObject * gobject)
   if (self->dmabuf)
     zwp_linux_dmabuf_v1_destroy (self->dmabuf);
 
+  if (self->gbm)
+    gbm_buffer_backend_destroy(self->gbm);
+
   if (self->shell)
     wl_shell_destroy (self->shell);
 
@@ -113,6 +120,11 @@ gst_wl_display_finalize (GObject * gobject)
     wl_display_disconnect (self->display);
   }
 
+  if (self->gbmfd) {
+    gbm_device_destroy(self->gbmdevice);
+    close(self->gbmfd);
+  }
+
   G_OBJECT_CLASS (gst_wl_display_parent_class)->finalize (gobject);
 }
 
@@ -148,7 +160,7 @@ gst_wl_display_roundtrip (GstWlDisplay * self)
 }
 
 static void
-shm_format (void *data, struct wl_shm *wl_shm, guint format)
+shm_format (void *data, struct wl_shm *wl_shm, uint32_t format)
 {
   GstWlDisplay *self = data;
 
@@ -161,7 +173,7 @@ static const struct wl_shm_listener shm_listener = {
 
 static void
 dmabuf_format (void *data, struct zwp_linux_dmabuf_v1 *zwp_linux_dmabuf,
-    guint format)
+    uint32_t format)
 {
   GstWlDisplay *self = data;
 
@@ -217,9 +229,32 @@ gst_wl_display_check_format_for_dmabuf (GstWlDisplay * display,
   return FALSE;
 }
 
+gboolean
+gst_wl_display_check_format_for_gbm (GstWlDisplay * display,
+    GstVideoFormat format)
+{
+  GArray *formats;
+  guint i, gbm_fmt;
+
+  if (!display->gbm)
+    return FALSE;
+
+  gbm_fmt = gst_video_format_to_wl_gbm_format (format);
+  if (gbm_fmt == (guint) - 1)
+    return FALSE;
+
+  formats = display->gbm_formats;
+  for (i = 0; i < formats->len; i++) {
+    if (g_array_index (formats, guint, i) == gbm_fmt)
+      return TRUE;
+  }
+
+  return FALSE;
+}
+
 static void
 registry_handle_global (void *data, struct wl_registry *registry,
-    guint id, const char *interface, guint version)
+    uint32_t id, const char *interface, uint32_t version)
 {
   GstWlDisplay *self = data;
 
@@ -236,6 +271,7 @@ registry_handle_global (void *data, struct wl_registry *registry,
     xdg_shell_use_unstable_version(self->xdgshell, 5);
   } else if (g_strcmp0 (interface, "wl_shm") == 0) {
     self->shm = wl_registry_bind (registry, id, &wl_shm_interface, 1);
+    shm_format(self, NULL, WL_SHM_FORMAT_NV12);
     wl_shm_add_listener (self->shm, &shm_listener, self);
   } else if (g_strcmp0 (interface, "wp_viewporter") == 0) {
     self->viewporter =
@@ -247,6 +283,23 @@ registry_handle_global (void *data, struct wl_registry *registry,
     self->dmabuf =
         wl_registry_bind (registry, id, &zwp_linux_dmabuf_v1_interface, 1);
     zwp_linux_dmabuf_v1_add_listener (self->dmabuf, &dmabuf_listener, self);
+  } else if (g_strcmp0 (interface, "gbm_buffer_backend") == 0) {
+    self->gbm =
+        wl_registry_bind(registry, id, &gbm_buffer_backend_interface, 1);
+
+    self->gbmfd = open("/dev/dri/card0", O_RDWR);
+    g_return_if_fail(self->gbmfd >= 0);
+
+    self->gbmdevice = gbm_create_device(self->gbmfd);
+    g_return_if_fail(self->gbmdevice != NULL);
+
+    guint formats[] = {
+        GBM_FORMAT_XRGB8888, GBM_FORMAT_ARGB8888, GBM_FORMAT_XBGR8888,
+        GBM_FORMAT_RGBX8888, GBM_FORMAT_ABGR8888, GBM_FORMAT_RGBA8888,
+        GBM_FORMAT_RGB888, GBM_FORMAT_RGB565, GBM_FORMAT_BGR565,
+        GBM_FORMAT_NV12, GBM_FORMAT_NV21_ZSL,
+    };
+    g_array_append_vals (self->gbm_formats, formats, G_N_ELEMENTS (formats));
   }
 }
 
@@ -370,6 +423,10 @@ gst_wl_display_new_existing (struct wl_display * display,
     g_warning ("Could not bind to zwp_linux_dmabuf_v1");
   }
 
+  if (!self->gbm) {
+    g_warning ("Could not bind to gbm_buffer_backend");
+  }
+
   self->thread = g_thread_try_new ("GstWlDisplay", gst_wl_display_thread_run,
       self, &err);
   if (err) {
diff --git a/ext/wayland/wldisplay.h b/ext/wayland/wldisplay.h
index cd061cd..accc96a 100644
--- a/ext/wayland/wldisplay.h
+++ b/ext/wayland/wldisplay.h
@@ -23,11 +23,13 @@
 
 #include <gst/gst.h>
 #include <gst/video/video.h>
+#include <gbm_priv.h>
 #include <wayland-client.h>
 #include "xdg-shell-client-protocol.h"
 #include "viewporter-client-protocol.h"
 #include "scaler-client-protocol.h"
 #include "linux-dmabuf-unstable-v1-client-protocol.h"
+#include "gbm-buffer-backend-client-protocol.h"
 
 G_BEGIN_DECLS
 
@@ -59,14 +61,20 @@ struct _GstWlDisplay
   struct wp_viewporter *viewporter;
   struct wl_scaler *scaler;
   struct zwp_linux_dmabuf_v1 *dmabuf;
+  struct gbm_buffer_backend *gbm;
   GArray *shm_formats;
   GArray *dmabuf_formats;
+  GArray *gbm_formats;
+
+  struct gbm_device *gbmdevice;
 
   /* private */
   gboolean own_display;
   GThread *thread;
   GstPoll *wl_fd_poll;
 
+  gint gbmfd;
+
   GMutex buffers_mutex;
   GHashTable *buffers;
   gboolean shutting_down;
@@ -91,6 +99,8 @@ gboolean gst_wl_display_check_format_for_shm (GstWlDisplay * display,
     GstVideoFormat format);
 gboolean gst_wl_display_check_format_for_dmabuf (GstWlDisplay * display,
     GstVideoFormat format);
+gboolean gst_wl_display_check_format_for_gbm (GstWlDisplay * display,
+    GstVideoFormat format);
 
 G_END_DECLS
 
diff --git a/ext/wayland/wlgbmallocator.c b/ext/wayland/wlgbmallocator.c
new file mode 100644
index 0000000..6ed90ea
--- /dev/null
+++ b/ext/wayland/wlgbmallocator.c
@@ -0,0 +1,159 @@
+/* GStreamer Wayland video sink
+ *
+ * Copyright (C) 2016 STMicroelectronics SA
+ * Copyright (C) 2016 Fabien Dessenne <fabien.dessenne@st.com>
+ *
+ * This library is free software; you can redistribute it and/or
+ * modify it under the terms of the GNU Library General Public
+ * License as published by the Free Software Foundation; either
+ * version 2 of the License, or (at your option) any later version.
+ *
+ * This library is distributed in the hope that it will be useful,
+ * but WITHOUT ANY WARRANTY; without even the implied warranty of
+ * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the GNU
+ * Library General Public License for more details.
+ *
+ * You should have received a copy of the GNU Library General Public
+ * License along with this library; if not, write to the Free
+ * Software Foundation, Inc., 51 Franklin Street, Fifth Floor,
+ * Boston, MA 02110-1301 USA.
+ */
+
+#ifdef HAVE_CONFIG_H
+#include <config.h>
+#endif
+
+#include <gst/allocators/gstfdmemory.h>
+
+#include "wlgbmallocator.h"
+#include "wlvideoformat.h"
+
+GST_DEBUG_CATEGORY_EXTERN (gstwayland_debug);
+#define GST_CAT_DEFAULT gstwayland_debug
+
+typedef struct
+{
+  GMutex lock;
+  struct wl_buffer *wbuffer;
+  GCond cond;
+} GbmBufferData;
+
+static void
+gbm_create_succeeded(void *data, struct gbm_buffer_params *params,
+    struct wl_buffer *wbuffer)
+{
+  GbmBufferData *gbmdata = (GbmBufferData *) data;
+
+  g_mutex_lock (&gbmdata->lock);
+
+  gbmdata->wbuffer = wbuffer;
+  gbm_buffer_params_destroy (params);
+  g_cond_signal (&gbmdata->cond);
+
+  g_mutex_unlock (&gbmdata->lock);
+}
+
+static void
+gbm_create_failed(void *data, struct gbm_buffer_params *params)
+{
+  GbmBufferData *gbmdata = (GbmBufferData *) data;
+
+  g_mutex_lock (&gbmdata->lock);
+
+  gbmdata->wbuffer = NULL;
+  gbm_buffer_params_destroy (params);
+  g_cond_signal (&gbmdata->cond);
+
+  g_mutex_unlock (&gbmdata->lock);
+}
+
+static const struct gbm_buffer_params_listener gbm_params_listener = {
+    gbm_create_succeeded,
+    gbm_create_failed
+};
+
+struct wl_buffer *
+gst_wl_gbm_memory_construct_wl_buffer (GstBuffer * buffer,
+    GstWlDisplay * display, const GstVideoInfo * info)
+{
+  gint fd, metafd, format, flags = 0;
+  guint width, height;
+  struct gbm_buffer_params *params;
+
+  struct gbm_import_fd_data bufinfo;
+  struct gbm_bo *bo;
+
+  GbmBufferData gbmdata;
+  gint64 timeout;
+
+  g_return_val_if_fail (gst_wl_display_check_format_for_gbm (display,
+      GST_VIDEO_INFO_FORMAT (info)), NULL);
+
+  fd = gst_fd_memory_get_fd(gst_buffer_peek_memory (buffer, 0));
+  format = gst_video_format_to_wl_gbm_format (GST_VIDEO_INFO_FORMAT (info));
+
+  width = GST_VIDEO_INFO_WIDTH (info);
+  height = GST_VIDEO_INFO_HEIGHT (info);
+
+  bufinfo.fd = fd;
+  bufinfo.width = width;
+  bufinfo.height = height;
+  bufinfo.format = format;
+
+  bo = gbm_bo_import (display->gbmdevice, GBM_BO_IMPORT_FD,
+                      &bufinfo, GBM_BO_USE_SCANOUT);
+  g_return_val_if_fail (bo != NULL, NULL);
+
+  gbm_perform (GBM_PERFORM_GET_METADATA_ION_FD, bo, &metafd);
+  g_return_val_if_fail (metafd != -1, NULL);
+
+  GST_DEBUG_OBJECT (display, "Creating wl_buffer from GBM of size %"
+      G_GSSIZE_FORMAT " fd %d, metafd %d, (%d x %d), format %s", info->size,
+      fd, metafd, width, height, gst_wl_gbm_format_to_string (format));
+
+  if (GST_BUFFER_FLAG_IS_SET (buffer, GST_VIDEO_BUFFER_FLAG_INTERLACED)) {
+    GST_DEBUG_OBJECT (display, "interlaced buffer");
+    flags = GBM_BUFFER_PARAMS_FLAGS_INTERLACED;
+
+    if (!GST_BUFFER_FLAG_IS_SET (buffer, GST_VIDEO_BUFFER_FLAG_TFF)) {
+      GST_DEBUG_OBJECT (display, "with bottom field first");
+      flags |= GBM_BUFFER_PARAMS_FLAGS_BOTTOM_FIRST;
+    }
+  }
+
+  g_cond_init (&gbmdata.cond);
+  g_mutex_init (&gbmdata.lock);
+  g_mutex_lock (&gbmdata.lock);
+
+  params = gbm_buffer_backend_create_params (display->gbm);
+
+  /* Request buffer creation */
+  gbm_buffer_params_add_listener (params, &gbm_params_listener, &gbmdata);
+  gbm_buffer_params_create (params, fd, metafd, width, height, format, flags);
+
+  /* Wait for the request answer */
+  wl_display_flush (display->display);
+
+  gbmdata.wbuffer = NULL;
+  timeout = g_get_monotonic_time () + G_TIME_SPAN_SECOND;
+
+  if (!g_cond_wait_until (&gbmdata.cond, &gbmdata.lock, timeout)) {
+    GST_ERROR_OBJECT (display, "GBM buffer creation timed out");
+    gbm_buffer_params_destroy (params);
+  }
+
+  if (!gbmdata.wbuffer) {
+    GST_ERROR_OBJECT (display, "Can't create GBM wl_buffer!");
+  } else {
+    GST_DEBUG_OBJECT (display, "Created GBM wl_buffer (%p): %dx%d, fmt=%.4s",
+        gbmdata.wbuffer, width, height, (char *) &format);
+  }
+
+  g_mutex_unlock (&gbmdata.lock);
+  g_mutex_clear (&gbmdata.lock);
+  g_cond_clear (&gbmdata.cond);
+
+  gbm_bo_destroy (bo);
+
+  return gbmdata.wbuffer;
+}
diff --git a/ext/wayland/wlgbmallocator.h b/ext/wayland/wlgbmallocator.h
new file mode 100644
index 0000000..87d127b
--- /dev/null
+++ b/ext/wayland/wlgbmallocator.h
@@ -0,0 +1,38 @@
+/* GStreamer Wayland video sink
+ *
+ * Copyright (C) 2016 STMicroelectronics SA
+ * Copyright (C) 2016 Fabien Dessenne <fabien.dessenne@st.com>
+ *
+ * This library is free software; you can redistribute it and/or
+ * modify it under the terms of the GNU Library General Public
+ * License as published by the Free Software Foundation; either
+ * version 2 of the License, or (at your option) any later version.
+ *
+ * This library is distributed in the hope that it will be useful,
+ * but WITHOUT ANY WARRANTY; without even the implied warranty of
+ * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the GNU
+ * Library General Public License for more details.
+ *
+ * You should have received a copy of the GNU Library General Public
+ * License along with this library; if not, write to the Free
+ * Software Foundation, Inc., 51 Franklin Street, Fifth Floor,
+ * Boston, MA 02110-1301 USA.
+ */
+
+#ifndef __GST_WL_GBM_ALLOCATOR_H__
+#define __GST_WL_GBM_ALLOCATOR_H__
+
+#include "gstwaylandsink.h"
+
+G_BEGIN_DECLS
+
+#ifndef GST_CAPS_FEATURE_MEMORY_GBM
+#define GST_CAPS_FEATURE_MEMORY_GBM "memory:GBM"
+#endif
+
+struct wl_buffer * gst_wl_gbm_memory_construct_wl_buffer (GstBuffer * buffer,
+    GstWlDisplay * display, const GstVideoInfo * info);
+
+G_END_DECLS
+
+#endif /* __GST_WL_GBM_ALLOCATOR_H__ */
diff --git a/ext/wayland/wlvideoformat.c b/ext/wayland/wlvideoformat.c
index 77cfac8..653a52c 100644
--- a/ext/wayland/wlvideoformat.c
+++ b/ext/wayland/wlvideoformat.c
@@ -42,6 +42,12 @@ typedef struct
   GstVideoFormat gst_format;
 } wl_DmabufVideoFormat;
 
+typedef struct
+{
+  guint wl_format;
+  GstVideoFormat gst_format;
+} wl_GbmVideoFormat;
+
 static const wl_ShmVideoFormat shm_formats[] = {
 #if G_BYTE_ORDER == G_BIG_ENDIAN
   {WL_SHM_FORMAT_XRGB8888, GST_VIDEO_FORMAT_xRGB},
@@ -85,31 +91,57 @@ static const wl_ShmVideoFormat shm_formats[] = {
 
 static const wl_DmabufVideoFormat dmabuf_formats[] = {
 #if G_BYTE_ORDER == G_BIG_ENDIAN
-  {DRM_FORMAT_XRGB8888, GST_VIDEO_FORMAT_xRGB},
-  {DRM_FORMAT_ARGB8888, GST_VIDEO_FORMAT_ARGB},
-  {DRM_FORMAT_XBGR8888, GST_VIDEO_FORMAT_xBGR},
-  {DRM_FORMAT_RGBX8888, GST_VIDEO_FORMAT_RGBx},
-  {DRM_FORMAT_BGRX8888, GST_VIDEO_FORMAT_BGRx},
-  {DRM_FORMAT_ABGR8888, GST_VIDEO_FORMAT_ABGR},
-  {DRM_FORMAT_RGBA8888, GST_VIDEO_FORMAT_RGBA},
-  {DRM_FORMAT_BGRA8888, GST_VIDEO_FORMAT_BGRA},
-  {DRM_FORMAT_RGB888, GST_VIDEO_FORMAT_RGB},
-  {DRM_FORMAT_BGR888, GST_VIDEO_FORMAT_BGR},
+  {GBM_FORMAT_XRGB8888, GST_VIDEO_FORMAT_xRGB},
+  {GBM_FORMAT_ARGB8888, GST_VIDEO_FORMAT_ARGB},
+  {GBM_FORMAT_XBGR8888, GST_VIDEO_FORMAT_xBGR},
+  {GBM_FORMAT_RGBX8888, GST_VIDEO_FORMAT_RGBx},
+  {GBM_FORMAT_BGRX8888, GST_VIDEO_FORMAT_BGRx},
+  {GBM_FORMAT_ABGR8888, GST_VIDEO_FORMAT_ABGR},
+  {GBM_FORMAT_RGBA8888, GST_VIDEO_FORMAT_RGBA},
+  {GBM_FORMAT_BGRA8888, GST_VIDEO_FORMAT_BGRA},
+  {GBM_FORMAT_RGB888, GST_VIDEO_FORMAT_RGB},
+  {GBM_FORMAT_BGR888, GST_VIDEO_FORMAT_BGR},
+#else
+  {GBM_FORMAT_XRGB8888, GST_VIDEO_FORMAT_BGRx},
+  {GBM_FORMAT_ARGB8888, GST_VIDEO_FORMAT_BGRA},
+  {GBM_FORMAT_XBGR8888, GST_VIDEO_FORMAT_RGBx},
+  {GBM_FORMAT_RGBX8888, GST_VIDEO_FORMAT_xBGR},
+  {GBM_FORMAT_BGRX8888, GST_VIDEO_FORMAT_xRGB},
+  {GBM_FORMAT_ABGR8888, GST_VIDEO_FORMAT_RGBA},
+  {GBM_FORMAT_RGBA8888, GST_VIDEO_FORMAT_ABGR},
+  {GBM_FORMAT_BGRA8888, GST_VIDEO_FORMAT_ARGB},
+  {GBM_FORMAT_RGB888, GST_VIDEO_FORMAT_BGR},
+  {GBM_FORMAT_BGR888, GST_VIDEO_FORMAT_RGB},
+#endif
+  {GBM_FORMAT_RGB565, GST_VIDEO_FORMAT_RGB16},
+  {GBM_FORMAT_YUYV, GST_VIDEO_FORMAT_YUY2},
+  {GBM_FORMAT_NV12, GST_VIDEO_FORMAT_NV12},
+};
+
+static const wl_GbmVideoFormat gbm_formats[] = {
+#if G_BYTE_ORDER == G_BIG_ENDIAN
+  {GBM_FORMAT_XRGB8888, GST_VIDEO_FORMAT_xRGB},
+  {GBM_FORMAT_ARGB8888, GST_VIDEO_FORMAT_ARGB},
+  {GBM_FORMAT_XBGR8888, GST_VIDEO_FORMAT_xBGR},
+  {GBM_FORMAT_RGBX8888, GST_VIDEO_FORMAT_RGBx},
+  {GBM_FORMAT_ABGR8888, GST_VIDEO_FORMAT_ABGR},
+  {GBM_FORMAT_RGBA8888, GST_VIDEO_FORMAT_RGBA},
+  {GBM_FORMAT_RGB888, GST_VIDEO_FORMAT_RGB},
+  {GBM_FORMAT_RGB565, GST_VIDEO_FORMAT_RGB16},
+  {GBM_FORMAT_BGR565, GST_VIDEO_FORMAT_BGR16},
 #else
-  {DRM_FORMAT_XRGB8888, GST_VIDEO_FORMAT_BGRx},
-  {DRM_FORMAT_ARGB8888, GST_VIDEO_FORMAT_BGRA},
-  {DRM_FORMAT_XBGR8888, GST_VIDEO_FORMAT_RGBx},
-  {DRM_FORMAT_RGBX8888, GST_VIDEO_FORMAT_xBGR},
-  {DRM_FORMAT_BGRX8888, GST_VIDEO_FORMAT_xRGB},
-  {DRM_FORMAT_ABGR8888, GST_VIDEO_FORMAT_RGBA},
-  {DRM_FORMAT_RGBA8888, GST_VIDEO_FORMAT_ABGR},
-  {DRM_FORMAT_BGRA8888, GST_VIDEO_FORMAT_ARGB},
-  {DRM_FORMAT_RGB888, GST_VIDEO_FORMAT_BGR},
-  {DRM_FORMAT_BGR888, GST_VIDEO_FORMAT_RGB},
+  {GBM_FORMAT_XRGB8888, GST_VIDEO_FORMAT_BGRx},
+  {GBM_FORMAT_ARGB8888, GST_VIDEO_FORMAT_BGRA},
+  {GBM_FORMAT_XBGR8888, GST_VIDEO_FORMAT_RGBx},
+  {GBM_FORMAT_RGBX8888, GST_VIDEO_FORMAT_xBGR},
+  {GBM_FORMAT_ABGR8888, GST_VIDEO_FORMAT_RGBA},
+  {GBM_FORMAT_RGBA8888, GST_VIDEO_FORMAT_ABGR},
+  {GBM_FORMAT_RGB888, GST_VIDEO_FORMAT_BGR},
+  {GBM_FORMAT_RGB565, GST_VIDEO_FORMAT_BGR16},
+  {GBM_FORMAT_BGR565, GST_VIDEO_FORMAT_RGB16},
 #endif
-  {DRM_FORMAT_RGB565, GST_VIDEO_FORMAT_RGB16},
-  {DRM_FORMAT_YUYV, GST_VIDEO_FORMAT_YUY2},
-  {DRM_FORMAT_NV12, GST_VIDEO_FORMAT_NV12},
+  {GBM_FORMAT_NV12, GST_VIDEO_FORMAT_NV12},
+  {GBM_FORMAT_NV21_ZSL, GST_VIDEO_FORMAT_NV21},
 };
 
 enum wl_shm_format
@@ -138,6 +170,19 @@ gst_video_format_to_wl_dmabuf_format (GstVideoFormat format)
   return -1;
 }
 
+gint
+gst_video_format_to_wl_gbm_format (GstVideoFormat format)
+{
+  guint i;
+
+  for (i = 0; i < G_N_ELEMENTS (gbm_formats); i++)
+    if (gbm_formats[i].gst_format == format)
+      return gbm_formats[i].wl_format;
+
+  GST_WARNING ("wayland gbm video format not found");
+  return -1;
+}
+
 GstVideoFormat
 gst_wl_shm_format_to_video_format (enum wl_shm_format wl_format)
 {
@@ -162,6 +207,18 @@ gst_wl_dmabuf_format_to_video_format (guint wl_format)
   return GST_VIDEO_FORMAT_UNKNOWN;
 }
 
+GstVideoFormat
+gst_wl_gbm_format_to_video_format (guint wl_format)
+{
+  guint i;
+
+  for (i = 0; i < G_N_ELEMENTS (gbm_formats); i++)
+    if (gbm_formats[i].wl_format == wl_format)
+      return gbm_formats[i].gst_format;
+
+  return GST_VIDEO_FORMAT_UNKNOWN;
+}
+
 const gchar *
 gst_wl_shm_format_to_string (enum wl_shm_format wl_format)
 {
@@ -175,3 +232,10 @@ gst_wl_dmabuf_format_to_string (guint wl_format)
   return gst_video_format_to_string
       (gst_wl_dmabuf_format_to_video_format (wl_format));
 }
+
+const gchar *
+gst_wl_gbm_format_to_string (guint wl_format)
+{
+  return gst_video_format_to_string
+      (gst_wl_gbm_format_to_video_format (wl_format));
+}
diff --git a/ext/wayland/wlvideoformat.h b/ext/wayland/wlvideoformat.h
index 331f582..4c2b815 100644
--- a/ext/wayland/wlvideoformat.h
+++ b/ext/wayland/wlvideoformat.h
@@ -26,16 +26,19 @@
 
 #include <wayland-client-protocol.h>
 #include <gst/video/video.h>
-#include <drm_fourcc.h>
+#include <gbm_priv.h>
 
 G_BEGIN_DECLS
 
 enum wl_shm_format gst_video_format_to_wl_shm_format (GstVideoFormat format);
 gint gst_video_format_to_wl_dmabuf_format (GstVideoFormat format);
+gint gst_video_format_to_wl_gbm_format (GstVideoFormat format);
 GstVideoFormat gst_wl_shm_format_to_video_format (enum wl_shm_format wl_format);
 GstVideoFormat gst_wl_dmabuf_format_to_video_format (guint wl_format);
+GstVideoFormat gst_wl_gbm_format_to_video_format (guint wl_format);
 const gchar *gst_wl_shm_format_to_string (enum wl_shm_format wl_format);
 const gchar *gst_wl_dmabuf_format_to_string (guint wl_format);
+const gchar *gst_wl_gbm_format_to_string (guint wl_format);
 
 G_END_DECLS
 
-- 
1.9.1

